### PR TITLE
[MRG+1] A canonical brain mask

### PR DESCRIPTION
**Pull request by KamalakerDadi**
As imposed in issue #945 created using MNI template.

Output of the **brain mask** with selected coordinates:
brain

### CODE REVIEW

File: nilearn/datasets/struct.py [hunk 1]
```diff
@@ -118,6 +119,25 @@ def load_mni152_template():
     return check_niimg(path)
 
 
+def load_brain_mask():
```
Comment (AlexandreAbraham) [inline nilearn/datasets/struct.py]: Maybe `load_mni152_brain_mask` would be a better naming?

Comment (KamalakerDadi) [inline nilearn/datasets/struct.py]: Ok.

Comment (bthirion) [inline nilearn/datasets/struct.py]: +1

**Comment by GaelVaroquaux** [comment]
Is this an MRG?

**Comment by KamalakerDadi** [comment]
I am pushing new commit which contains grey matter mask. Just about to push.

**Comment by KamalakerDadi** [comment]
@GaelVaroquaux please could you check ? How it goes ?


gery

File: nilearn/datasets/struct.py [hunk 2]
```diff
@@ -118,6 +120,75 @@ def load_mni152_template():
...
+    target_img = load_mni152_template()
+    mask_voxels = (target_img.get_data() > 0).astype(int)
+    mask_img = new_img_like(target_img, mask_voxels)
+    return mask_img
+
+
+def fetch_mni152_grey_matter_mask(data_dir=None, resume=True, verbose=1):
+    """Downloads ICBM152 template first, then loads computed 'gm' mask image.
+
+    The output 'gm' mask image will be resampled to MNI152 template space.
```
Comment (GaelVaroquaux) [inline nilearn/datasets/struct.py]: Why are you resampling? I would keep it as such.

Comment (KamalakerDadi) [inline nilearn/datasets/struct.py]: >Why are you resampling?

What if somebody want to use in their own experiments ?

Comment (GaelVaroquaux) [inline nilearn/datasets/struct.py]: > What if somebody want to use in their own experiments ?

They resample. The nifti masker does that automatically. No need to
resample twice.

Comment (KamalakerDadi) [inline nilearn/datasets/struct.py]: >They resample. The nifti masker does that automatically. No need to resample twice.

Ok, do you think I am over enthusiastic ? I don't think then it is necessary ```fetch_mni152_grey_matter_mask``` ?

Comment (GaelVaroquaux) [inline nilearn/datasets/struct.py]: > I don't think then it is necessary fetch_mni152_grey_matter_mask ?

I don't understand this sentence.

File: nilearn/datasets/struct.py [hunk 3]
```diff
@@ -118,6 +120,75 @@ def load_mni152_template():
...
+    icbm = fetch_icbm152_2009(data_dir=data_dir, resume=resume, verbose=verbose)
+    icbm_gm = icbm['gm']
+
+    brain_mask = load_mni152_brain_mask()
+    # resampling grey matter mask from ICBM to MNI template
+    gm_resampled_img = resample_to_img(icbm_gm, brain_mask)
+
+    gm_map = gm_resampled_img.get_data()
+    # getting one fifth of the values
+    gm_mask = (gm_map > 0.2)
```
Comment (GaelVaroquaux) [inline nilearn/datasets/struct.py]: Why aren't we using math_img here? I would image that it is a good fit.

Comment (GaelVaroquaux) [inline nilearn/datasets/struct.py]: Nope, sorry, stupid comment. Forget that.

File: nilearn/datasets/struct.py (hunk 2 above)
Comment (KamalakerDadi) [inline nilearn/datasets/struct.py]: Ok, I return grey mask as it is without resampling.

**Comment by GaelVaroquaux** [comment]
Could you add some seealso in the docstring of the new functions, and maybe the load_mni152

File: nilearn/datasets/struct.py [hunk 4]
```diff
@@ -118,6 +120,75 @@ def load_mni152_template():
...
+
+    See Also
+    --------
+    nilearn.datasets.fetch_icbm152_2009: for details regarding the ICBM152
+        template.
+
+    nilearn.datasets.load_mni152_template: for details regarding MNI152 template.
+
+    nilearn.image.resample_to_img: to resample source image to target image.
+    """
```
Comment (GaelVaroquaux) [inline nilearn/datasets/struct.py]: You should add a notes section that tell us that you threshold the ICBM probabilities at .2

**Comment by GaelVaroquaux** [comment]
Also, there needs entries added in the modules/reference.rst for the reference documentation.

File: nilearn/datasets/struct.py (hunk 4 above)
Comment (AlexandreAbraham) [inline nilearn/datasets/struct.py]: The mask seems very permissive, why .2? .5 seems more logical to me.

Comment (GaelVaroquaux) [inline nilearn/datasets/struct.py]: > The mask seems very permissive, why .2? .5 seems more logical to me.

Let's make it an optional argument. Permissive masks are usual a good
idea, however, so I think that .2 is a good default.

Comment (KamalakerDadi) [inline nilearn/datasets/struct.py]: >so I think that .2 is a good default.

For me too, for completeness I would post some pictures between those two values.

Comment (KamalakerDadi) [inline nilearn/datasets/struct.py]: >so I think that .2 is a good default.

For me too, for completeness I would post some pictures between those two values.

**Comment by KamalakerDadi** [comment]
So far grey matter output:
grey_comparison

@GaelVaroquaux Any quick advice how to complete holes ?

Side: threshold=0.5 or 0.2 doesn't show any difference

**Comment by GaelVaroquaux** [comment]
I don't see the holes as been a problem. It would be interesting to check whether they are actually holes in 3D, or just seen as such on the cuts.

**Comment by KamalakerDadi** [comment]
I checked seen as cuts.

File: nilearn/datasets/struct.py [hunk 5]
```diff
@@ -118,6 +120,91 @@ def load_mni152_template():
...
+        template and related.
+    """
+    # Load MNI template
+    target_img = load_mni152_template()
+    mask_voxels = (target_img.get_data() > 0).astype(int)
+    mask_img = new_img_like(target_img, mask_voxels)
+    return mask_img
+
+
+def fetch_brain_gm_mask(data_dir=None, threshold=0.2, resume=True, verbose=1):
```
Comment (AlexandreAbraham) [inline nilearn/datasets/struct.py]: I would add the name of the dataset in the function name.

File: nilearn/datasets/struct.py [hunk 6]
```diff
@@ -118,6 +120,91 @@ def load_mni152_template():
...
+    # Fetching ICBM152 grey matter mask image
+    icbm = fetch_icbm152_2009(data_dir=data_dir, resume=resume, verbose=verbose)
+    gm = icbm['gm']
+    gm_img = check_niimg(gm)
+    gm_data = niimg._safe_get_data(gm_img)
+
+    # getting one fifth of the values
+    gm_mask = (gm_data > threshold)
+
+    gm_mask = ndimage.binary_closing(gm_mask, iterations=2)
```
Comment (AlexandreAbraham) [inline nilearn/datasets/struct.py]: I understand why I found your mask thick. Why do you apply binary closing?

Comment (KamalakerDadi) [inline nilearn/datasets/struct.py]: Just to cover missing areas, I suppose.

why_bin_close

Comment (KamalakerDadi) [inline nilearn/datasets/struct.py]: @AlexandreAbraham does that sounds convincing or not ?

File: doc/modules/reference.rst [hunk 7]
```diff
@@ -85,6 +85,8 @@ uses.
    fetch_megatrawls_netmats
    fetch_cobre
    load_mni152_template
+   load_mni152_brain_mask
+   fetch_brain_gm_mask
```
Comment (GaelVaroquaux) [inline doc/modules/reference.rst]: I think that I would prefer that file to be in alphabetic order.

File: nilearn/datasets/struct.py [hunk 8]
```diff
@@ -118,6 +120,91 @@ def load_mni152_template():
...
+
+    References
+    ----------
+    Refer to load_mni152_template function for more information about MNI
+    template
+
+    See Also
+    --------
+    nilearn.datasets.load_mni152_template for details about version of MNI152
+        template and related.
```
Comment (GaelVaroquaux) [inline nilearn/datasets/struct.py]: I would say here "for the MNI152 T1 template".

**Comment by GaelVaroquaux** [comment]
LGTM aside my 2 minor comments.

**Comment by GaelVaroquaux** [comment]
> Just a curious question: Do you think we need an example showing how to
> analyze this mask in their own data processing ? If Yes, how do I start
> ?

This is going to be used in the neurovault PR (which we need to
ressucitate).

**Pull request merged**